kong/plugins/ddtrace/span.lua

```diff
@@ -18,33 +18,57 @@ local span_mt = {
 local uint64_t = ffi.typeof("uint64_t")
 local int64_t = ffi.typeof("int64_t")
 
-local function generate_span_id()
+local function random_64bit()
     local x = rand_bytes(8)
     local b = {1ULL * byte(x,1), 1ULL * byte(x,2), 1ULL * byte(x,3), 1ULL * byte(x,4), 1ULL * byte(x,5), 1ULL * byte(x,6), 1ULL * byte(x,7), 1ULL * byte(x,8)}
     local id = bit.bor(bit.lshift(b[1], 56), bit.lshift(b[2], 48), bit.lshift(b[3], 40), bit.lshift(b[4], 32), bit.lshift(b[5], 24), bit.lshift(b[6], 16), bit.lshift(b[7], 8), b[8])
-    -- truncate to 63-bit value
-    return bit.band(id, 0x7FFFFFFFFFFFFFFFULL)
+    return id
+end
+
+local function legacy_63bit_ids_generator()
+    -- NOTE(@dmehala): truncate to 63-bit value for legacy reasons
+    return bit.band(random_64bit(), 0x7FFFFFFFFFFFFFFFULL)
+end
+
+local function trace_64bit_ids_generator(_)
+    return {high=nil, low=legacy_63bit_ids_generator()}
+end
+
+local function trace_128bit_ids_generator(now_us)
+    local now_s = uint64_t(now_us / 1000000)
```

Comment:
I haven't dug into it, but perhaps you did: Make sure that this division makes sense in terms of types. I don't know if `now_us` (which I presume is a count of microseconds) is Lua's default number type, i.e. `double`, or if it's something like `uint64_t`. I'm curious to know which `operator /`, exactly, is happening here.

The distinction probably wouldn't change the answer enough to matter, but look into it if you haven't already.

Reply:
Ah, if it's a `uint64_t`, then this is truncating division, so fine.

If it's a `double`, then integer operations are exact up to `2**53 - 1`, which is is a date in 2255, so also fine.

If it's a `uint64_t` that gets coerced into a `double` before the division, then that's still fine for the same reason above.

Fine, then.